great_expectations/zep/interfaces.py

```diff
@@ -1,18 +1,64 @@
-from typing import List, Union
+import abc
+import dataclasses
+from typing import List, TypeVar, Generic, Optional
 
 from typing_extensions import Protocol
 
-from great_expectations.core.batch import Batch, BatchRequest, RuntimeBatchRequest
+from great_expectations.core.batch import BatchDataType
 from great_expectations.execution_engine import ExecutionEngine
 
 
+# The BatchRequestOptions provides extensibility for a specific execution engine.
+# A particular execution engine can subclass BatchRequestOptions and add its relevant properties.
+class BatchRequestOptions:
+    pass
+
+
+BatchRequestOptionsT = TypeVar("BatchRequestOptionsT", bound=BatchRequestOptions)
+
+
+# BDIRKS BDIRKS I think we should have the datasource and and data asset here
+#               We can define the names as properties.
+@dataclasses.dataclass(frozen=True)
+class BatchRequest(Generic[BatchRequestOptionsT]):
+    # This is a collection of immutable parameters used to get data from the datasource/data asset
+    # We plan to expand this to provide a mechanism to add request parameters that exist for all
+    # datasources.
+    datasource_name: str
+    data_asset_name: str
+    options: BatchRequestOptionsT
+
+    # Usage stats expects a method like this to exist but I'm not sure if we really want to
+    # implement it or not. Other changes would also be necessary. I've commented it out while
+    # I sort through the usage stats piece.
+    # def to_json_dict(self) -> dict:
+    #     # Convert the BatchRequest to a dictionary that is json serializable
+    #     # NOT a JSON string.
+    #     # This is necessary for usage stats.
+    #     return dataclasses.asdict(self)
+
+
+class DataAsset(abc.ABC, Generic[BatchRequestOptionsT]):
+    def __init__(self, name: str):
+        self._name = name
+
+    @property
+    def name(self):
+        return self._name
+
+    # BDIRKS not sure if quite right.
+    def get_batch_request(self, options: Optional[BatchRequestOptionsT]) -> BatchRequest[BatchRequestOptionsT]:
+        ...
+
+
 class Datasource(Protocol):
     execution_engine: ExecutionEngine
     asset_types: List[type]
+    name: str
 
     def get_batch_list_from_batch_request(
-        self, batch_request: Union[BatchRequest, RuntimeBatchRequest]
-    ) -> List[Batch]:
+            self, batch_request: "BatchRequest"
+    ) -> "List[Batch]":
```

Comment:
Should import `from __future__ annotations` here.

Reply:
Yes, this seems better.